warehouse/packaging/tasks.py

```diff
@@ -54,6 +60,77 @@ def sync_file_to_cache(request, file_id):
         file.cached = True
 
 
+@tasks.task(ignore_result=True, acks_late=True)
+def backfill_metadata(request):
+    metrics = request.find_service(IMetricsService, context=None)
+    base_url = request.registry.settings.get("files.url")
+
+    archive_storage = request.find_service(IFileStorage, name="archive")
+    cache_storage = request.find_service(IFileStorage, name="cache")
+    session = PipSession()
+
+    # Get all wheel files without metadata in reverse chronologicial order
+    files_without_metadata = (
+        request.db.query(File)
+        .filter(File.packagetype == "bdist_wheel")
+        .filter(File.metadata_file_sha256_digest.is_(None))
+        .order_by(desc(File.upload_time))
+    )
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        for file_ in files_without_metadata.yield_per(100):
+            # Use pip to download just the metadata of the wheel
+            file_url = base_url.format(path=file_.path)
+            lazy_dist = dist_from_wheel_url(
+                file_.release.project.normalized_name, file_url, session
+            )
+            wheel_metadata_contents = lazy_dist._dist._files[Path("METADATA")]
+
+            # Write the metadata to a temporary file
+            temporary_filename = os.path.join(tmpdir, file_.filename) + ".metadata"
+            with open(temporary_filename, "wb") as fp:
+                fp.write(wheel_metadata_contents)
+
+            # Hash the metadata and add it to the File instance
+            file_.metadata_file_sha256_digest = (
+                hashlib.sha256(wheel_metadata_contents).hexdigest().lower()
+            )
+            file_.metadata_file_blake2_256_digest = (
+                hashlib.blake2b(wheel_metadata_contents, digest_size=256 // 8)
+                .hexdigest()
+                .lower()
+            )
```

Comment:
I presume the digests here are written to the DB immediately? (Rather than when `_file`'s destructor is called)

If so, it seems like there is a failure mode here that could leave files without their metadata file?

ie: If the backfill task's worker process gets killed/restarted between writing the digests and storing the metadata file in the storage backend (or if the write to the storage backend fails). In such a case, the next time the backfill tasks runs, the wheel in question wouldn't be returned by the DB query since the digests would then be set - even though the metadata file was never uploaded.

If setting the digests was moved after writing to the storage backend, then in the case of an interrupted backfill task, the inconsistency between storage backend (which would have the uploaded file) and DB digest state (which wouldn't yet have the digest) would instead self-heal on the next run.

Reply:
this is a non-issue, the changes to the database aren't persisted until the entire request has completed and a transaction is committed.

Reply:
Ah good to know. (The fact that everything was configured to use transactions isn't obvious from code inspection of this single task.)